nomos-da/full-replication/src/lib.rs

```diff
@@ -142,45 +127,46 @@ pub struct CertificateVerificationParameters {
 }
 
 impl certificate::Certificate for Certificate {
-    type VerificationParameters = CertificateVerificationParameters;
-    type Signature = [u8; 32];
     type Id = [u8; 32];
-
-    fn signers(&self) -> Vec<bool> {
-        self.attestations.iter().map(|_| true).collect()
-    }
+    type Signature = [u8; 32];
+    type VerificationParameters = CertificateVerificationParameters;
 
     fn signature(&self) -> Self::Signature {
-        let signatures: Vec<u8> = self
-            .attestations
-            .iter()
-            .map(|attestation| attestation.attestation_signature())
-            .flatten()
-            .collect();
+        let mut attestations = self.attestations.clone();
+        attestations.sort();
+        let mut signatures = Vec::new();
+        for attestation in &attestations {
+            signatures.extend_from_slice(attestation.signature());
+        }
         hash(signatures)
     }
 
     fn id(&self) -> Self::Id {
         let mut input = self
             .attestations
             .iter()
-            .map(|a| a.attestation_signature())
+            .map(|a| a.signature())
             .collect::<Vec<_>>();
         // sort to make the hash deterministic
         input.sort();
         hash(input.concat())
     }
 
-    fn verify(&self, authorization_parameters: Self::VerificationParameters) -> bool {
-        authorization_parameters.threshold <= self.attestations.len()
+    fn signers(&self) -> Vec<bool> {
+        unimplemented!()
+    }
+
+    fn verify(&self, params: Self::VerificationParameters) -> bool {
+        self.attestations.len() >= params.threshold
     }
 }
 
-impl CertificateExtension for Certificate {
-    type Extension = ();
+impl metadata::Metadata for Certificate {
+    type AppId = [u8; 32];
+    type Index = u64;
 
-    fn extension(&self) -> Self::Extension {
-        ()
+    fn metadata(&self) -> Option<(Self::AppId, Self::Index)> {
+        Some((self.metadata.app_id, self.metadata.index))
```

Comment:
What happens if a certificate has no metadata? or when do this happens? Just trying to understand the `Option` in the trait 🤔

Reply:
I made it optional in the FullReplication implementation assuming that when the zone is receiving attestations, it might want to hold a partial certificate without metadata assigned. Verification on the node side should fail if metadata is not attached. We might improve this with the da-protocol-v1 certificate.

Reply:
I'm not sure that is the correct approach. I feel that if something do not have the metadata it should just simply not be able to implement the trait. But yeah, lets revisit this later on.